proposals/0370-modifiers.rst

```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
     have a known type if declared with a type signature. Alternatively, the
     modifier may have a top-level type signature.
 
-12. A modifier of type ``Multiplicity`` changes the multiplicity of the following arrow,
-    or following pattern-bound variable of a lambda,
-    or preceding record field.
-    Multiple modifiers of type ``Multiplicity`` on the same arrow are not allowed.
-    Any other use of a modifier is an error.
-
-13. ``-XLinearTypes`` implies ``-XModifiers``.
-
-14. Future modifiers will be put *before* the element they modify. Alternatively,
+12. Future modifiers will be put *before* the element they modify. Alternatively,
     a modifier can be put directly before a syntactic closer or separator, such
     as ``;`` or ``where`` or ``)``.
 
-15. Modifiers with an unknown meaning produce a warning, controlled by
-    ``-Wunknown-modifiers``. They are otherwise ignored. (However, in order to
-    know that a modifier is unknown, it still must be parsed, renamed, and type-checked.)
+13. Modifiers of unknown kind produce an error.
+
+14. Modifiers of known kind but with an unknown meaning produce a warning,
+    controlled by ``-Wunknown-modifiers``. They are otherwise ignored. (However,
+    in order to know that a modifier is unknown, it still must be parsed,
+    renamed, and type-checked.)
```

Comment:
This is not something that your changes here introduce, but how is "unknown meaning" defined?

Does this allow for a potential future where user provided modifiers exist?

Reply:
I guess something like, "the default behavior of modifiers is to emit this warning, but if GHC specifically recognizes a modifier in some context it overrides that behavior".

Not sure about user provided modifiers. There's speculation about handling them via plugin, and then maybe it would be possible for that plugin to mark them as "I've done something with this modifier, no need to warn about it"? But I don't know enough about plugins to know if that's possible.

Reply:
In my idealised world, which I believe I previously shared, if modifiers are available in the metadata for `GHC.Generics.Generic`, you would not need to do anything with plugins - you would simply be able to access them via the metadata in `Generic`, same as with any other metadata about types in `Generic`. 

This is more of an "open world" sort of deal, because now, downstream consumers of your library can see if you've put some modifiers on your type and then potentially consume them in some way, which, now that I've said that, kind of implies to me that this mode of operation (consider unknown-to-ghc modifiers to be "bad") conflicts with what I've been imagining.